### PR TITLE
Fix endian problem due to int/MonoThreadStateMachine confusion.

### PR DESCRIPTION
**Pull request by jaykrell**
Problems reported on AIX.
Fixes https://github.com/mono/mono/issues/17501.

### CODE REVIEW

**Comment by NattyNarwhal** [comment]
The initial version didn't change anything for me; new version you just pushed:

```
mono-threads.c: In function 'dump_threads':
mono-threads.c:299:71: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
   g_async_safe_printf ("--thread %p id %p [%p] state %x  %s\n", info, (void *) mono_thread_info_get_tid (info), (void*)(size_t)info->native_handle, info->thread_state, info == cur ? "GC INITIATOR" : "" );
                                                                       ^
mono-threads-state-machine.c: In function 'get_thread_suspend_count':
mono-threads-state-machine.c:29:34: error: incompatible types when initializing type 'unsigned int' using type 'MonoThreadStateMachine {aka union <anonymous>}'
  MonoThreadStateMachine state = {state};
                                  ^~~~~
mono-threads-state-machine.c: In function 'get_thread_no_safepoints':
mono-threads-state-machine.c:36:34: error: incompatible types when initializing type 'unsigned int' using type 'MonoThreadStateMachine {aka union <anonymous>}'
  MonoThreadStateMachine state = {state};
                                  ^~~~~
mono-threads-state-machine.c: In function 'thread_state_cas':
mono-threads-state-machine.c:60:30: warning: pointer targets in passing argument 1 of 'mono_atomic_cas_i32' differ in signedness [-Wpointer-sign]
  return mono_atomic_cas_i32 (&state->raw, new_raw, old_raw);
                              ^
In file included from ../../mono/utils/refcount.h:11:0,
                 from ../../mono/utils/mono-threads.h:23,
                 from mono-threads-state-machine.c:8:
../../mono/utils/atomic.h:242:22: note: expected 'volatile gint32 * {aka volatile int *}' but argument is of type 'guint * {aka unsigned int *}'
 static inline gint32 mono_atomic_cas_i32(volatile gint32 *dest,
                      ^~~~~~~~~~~~~~~~~~~
mono-threads-state-machine.c: In function 'unwrap_thread_state':
mono-threads-state-machine.c:92:36: warning: pointer targets in passing argument 1 of 'mono_atomic_load_i32' differ in signedness [-Wpointer-sign]
  state.raw = mono_atomic_load_i32 (&info->thread_state.raw);
                                    ^
In file included from ../../mono/utils/refcount.h:11:0,
                 from ../../mono/utils/mono-threads.h:23,
                 from mono-threads-state-machine.c:8:
../../mono/utils/atomic.h:304:22: note: expected 'volatile gint32 * {aka volatile int *}' but argument is of type 'guint * {aka unsigned int *}'
 static inline gint32 mono_atomic_load_i32(volatile gint32 *src)
                      ^~~~~~~~~~~~~~~~~~~~
At top level:
mono-threads-state-machine.c:34:1: warning: 'get_thread_no_safepoints' defined but not used [-Wunused-function]
 get_thread_no_safepoints (int thread_state)
 ^~~~~~~~~~~~~~~~~~~~~~~~
mono-threads-state-machine.c:27:1: warning: 'get_thread_suspend_count' defined but not used [-Wunused-function]
 get_thread_suspend_count (int thread_state)

```

(I lost DNS it seems, so I can't connect to Gitter...)

File: mono/utils/mono-threads-state-machine.c [hunk 1]
```diff
@@ -197,7 +194,7 @@ mono_threads_transition_attach (MonoThreadInfo* info)
 			mono_fatal_with_history ("suspend_count = %d, but should be == 0", suspend_count);
 		if (no_safepoints)
 			mono_fatal_with_history ("no_safepoints = TRUE, but should be FALSE");
-		if (thread_state_cas (&info->thread_state, STATE_RUNNING, raw_state) != raw_state)
+		if (thread_state_cas (&info->thread_state, build_thread_state (STATE_RUNNING, 0, 0), raw_state) != raw_state)
```
Comment (jaykrell) [inline mono/utils/mono-threads-state-machine.c]: These and similar are the importane lines.

File: mono/utils/mono-threads-state-machine.c [hunk 2]
```diff
@@ -18,49 +18,45 @@
...
 	state.state = thread_state;
 	state.no_safepoints = no_safepoints;
 	state.suspend_count = suspend_count;
-	return state.raw;
+	return state;
 }
 
 static int
-thread_state_cas (MonoThreadStateMachine *state, int new_raw, int old_raw)
+thread_state_cas (MonoThreadStateMachine *state, MonoThreadStateMachine new_value, int old_raw)
```
Comment (jaykrell) [inline mono/utils/mono-threads-state-machine.c]: This is a good change, because it prevents the wrong form from compiling -- typesafety.

**Comment by NattyNarwhal** [comment]
Seems to work for me. I had a stuck file, but probably just a fluke that triggered at a bad time.

File: mono/utils/mono-threads-state-machine.c (hunk 1 above)
Comment (lambdageek) [reply mono/utils/mono-threads-state-machine.c]: nit: the third argument to `build_thread_state` should be `FALSE`, not `0`.

File: mono/utils/mono-threads-state-machine.c [hunk 3]
```diff
@@ -18,49 +18,45 @@
 static int
 get_thread_state (int thread_state)
 {
-	MonoThreadStateMachine state;
-	state.raw = thread_state;
+	const MonoThreadStateMachine state = {thread_state};
```
Comment (lambdageek) [inline mono/utils/mono-threads-state-machine.c]: nit: this (and similar changes in the other helper functions) is an unrelated, unmotivated style change.

Comment (jaykrell) [reply mono/utils/mono-threads-state-machine.c]: A little shorter to read, but still clear.
A little more const.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Pull request merged**